### PR TITLE
code: add rejection sampling module (Chapter 9)

### PR DESCRIPTION
**Pull request by zafstojano**
## Summary

Adds an educational rejection sampling implementation for Chapter 9, matching the style of `policy_gradients/` and `direct_alignment/`.

Pipeline:
1. Generate N rollouts per GSM8k train prompt with `Qwen/Qwen3-1.7B`
2. Score every rollout with `nvidia/AceMath-7B-RM`
3. Select a subset via one of four strategies
4. SFT `Qwen/Qwen3-1.7B` on the selected subset
5. Evaluate exact-match accuracy on the GSM8k test split

The two chapter selection strategies (`top_per_prompt`, `top_k_overall`) are each paired with a matched random baseline at the same sample budget (`random_per_prompt`, `random_k_overall`) so comparing a pair isolates whether the reward model actually beats a coin flip. Stages 1+2 are cached to a single JSONL keyed by a hash of generation/scoring params, so all four strategies share the same rollout cache.

New console scripts in `pyproject.toml`:
- `rs-preprocess` — Stage 1 + Stage 2 only
- `rs-train` — full pipeline (auto-runs preprocess on cache miss)

## Results

GSM8k exact-match on the test slice (`max_test_samples: 200`), 1000 train prompts × 8 completions, 2 SFT epochs:

| strategy                      | test accuracy |
|-------------------------------|---------------|
| base `Qwen3-1.7B` (no SFT)    | 0.760         |
| `rs_random_per_prompt`        | 0.775         |
| `rs_top_k_overall`            | 0.785         |
| `rs_random_k_overall`         | 0.795         |
| `rs_top_per_prompt`           | **0.800**     |

`top_per_prompt` beats its random baseline as expected, and all four SFT variants clear the zero-SFT base model. Curiously, `top_k_overall` lands *below* its random baseline — inverted from the story the chapter tells. Implementation checks out (the selection module has a worked-example assertion from the chapter's reward matrix that still passes), so the most likely explanation is that this is an artifact of the very small train/eval slice used here as a toy example — with only 1000 train prompts and 200 eval questions, a 0.010 gap is well within run-to-run noise, and `top_k_overall` can concentrate many training pairs on a handful of easy prompts, hurting coverage relative to the uniform flat sample. Scaling train samples / eval set size would be needed to say anything stronger.

## Test plan

- [x] `python -m rejection_sampling.selection` — chapter worked-example (M=5, N=4) assertions pass for all four strategies
- [x] All four configs run end-to-end and log `test_accuracy` to wandb
- [x] `ruff check rejection_sampling/` — clean
- [x] Imports and module loads verified

Screenshot 2026-04-11 at 22 37 12

---

What is left as TODO is to run these experiments on your DGX spark, make sure they reproduce, and then update `code/README.md` with unified instructions and overview.

### CODE REVIEW

File: code/rejection_sampling/preprocess.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+    for i, prompt in enumerate(prompts):
+        for j, completion in enumerate(completions[i]):
+            chat_str = _build_scoring_chat(prompt["question"], completion, rm_tokenizer)
+            # Chat template already inserts special tokens; encoding with
+            # add_special_tokens=True would double-prepend them.
+            ids = rm_tokenizer.encode(chat_str, add_special_tokens=False)
+            flat.append((i, j, ids))
+    flat.sort(key=lambda item: len(item[2]))
+
+    pad_id = rm_tokenizer.pad_token_id or rm_tokenizer.eos_token_id
```
[the remark this answers is not in the document]
Comment (zafstojano) [reply code/rejection_sampling/preprocess.py]: Fixed in 5cfb7be. Swapped the `or` shortcut for an explicit `is not None` check at preprocess.py:168 and the two matching sites in train.py (lines 106 and 208) so a valid `pad_token_id == 0` no longer silently falls back to EOS. Note: in practice this is a latent bug rather than an active one — the default `nvidia/AceMath-7B-RM` has `pad_token_id is None` (so it hits the EOS branch either way), and `attention_mask` is zeroed at pad positions so a correctly-implemented seq-classification RM wouldn't attend to them regardless. But the fix is a one-liner and removes the footgun if someone swaps in a different RM, so worth doing.

File: code/rejection_sampling/utils.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def extract_gsm8k_answer(text: str) -> str | None:
+    """Pull a numeric answer out of a completion — last `\\boxed{...}` first, then last number."""
+    boxed_matches = _BOXED_RE.findall(text)
+    if boxed_matches:
+        inside = boxed_matches[-1]
+        numbers = _NUMBER_RE.findall(inside)
+        if numbers:
+            return numbers[-1].replace(",", "")
```
[the remark this answers is not in the document]
Comment (zafstojano) [reply code/rejection_sampling/utils.py:77]: Declining this one. The regex observation is correct — `_NUMBER_RE` on `1/2` returns `[1, 2]` and picks the last, so `\boxed{1/2}` → `"2"`, and `\boxed{\frac{3}{4}}` → `"4"`. But the bias-downward claim doesn't hold for how this extractor is actually used.

The only caller is train.py:253, which does strict exact-match against a GSM8k gold produced by `format_gsm8k_gold` — and GSM8k gold answers are **always integers**. So:

| Model output | Gold | Current | "Fixed" | Match? |
|---|---|---|---|---|
| `\boxed{1/2}` | `0.5` | `"2"` | `"1/2"` | Both wrong |
| `\boxed{12/2}` | `6` | `"2"` | `"12/2"` | Both wrong |
| `\boxed{\frac{3}{4}}` | `0.75` | `"4"` | `"\frac{3}{4}"` | Both wrong |

Preserving the raw fraction string doesn't increase `test_accuracy` against integer golds — the output is still the wrong string, just a different wrong string. (Amusingly, returning the last numeric token can even produce accidental *false positives*, e.g. `\boxed{4/2}` vs gold `2`.)

Making this actually improve accuracy would require a real math-equivalence check (SymPy / fraction-to-decimal), which is a separate design choice for a textbook example. Leaving the extractor as-is.

**Comment by zafstojano** [comment]
Disclaimer: the replies are from Claude Code :D

**Comment by natolambert** [comment]
> Disclaimer: the replies are from Claude Code :D

Claude needs a better feature for this. I use it a lot.
